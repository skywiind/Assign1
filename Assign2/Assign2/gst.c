/*****************************************************************************************
*                                           *                                            *
*             CS201-001 Assign 2            * This file defines the implementations of   *
*             Wyatt Fairbanks               * the GST class methods.                     *
*             Dr. John Lusth                *                                            *
*             gst.c                         *                                            *
*             10/22/2018                    *                                            *
*                                           *                                            *
*****************************************************************************************/
#include <stdio.h>
#include <stdlib.h>
#include "bst.h"
#include "gst.h"

struct gst {
	BST   *tree;
	void (*display) (void *value, FILE *fp);
	void (*free) (void *value);
	void (*swap) (TNODE *v1, TNODE *v2);
	int  (*compare) (void *v1, void *v2);
	int d;
};

typedef struct gnode GNODE;

struct gnode {
	void  *v;
	int    f;
	void (*display) (void *value, FILE *fp);
	void (*free) (void *value);
	int  (*compare) (void *v1, void *v2);
};

//UTILITY FUNCTIONS ////////////////////////////////////////////////////////////////

void displayGNODE(void *v, FILE *fp) {
	GNODE *temp = v;
	if (temp == NULL) {
		return;
	}
	temp->display(temp->v, fp);
	GNODE *test = temp->v;
	if (test->f > 1) {
		fprintf(fp, "<%d>", test->f);
	}
	/*
	if (temp->f > 1) {
		fprintf(fp, "<%d>", temp->f);
	}
	*/
	return;
}

int compareGNODE(void *v, void *w) {
	GNODE *a = v;
	GNODE *b = w;
	return b->compare(a->v, b->v);
}

void freeGNODE(void *n) {
	GNODE *temp = n;
	temp->free(temp->v);
	free(temp);
	return;
}

void swapper(TNODE *a, TNODE *b) {
	void *x = getTNODEvalue(a);
	void *y = getTNODEvalue(b);
	setTNODEvalue(a, y);
	setTNODEvalue(b, x);
	return;
}

//END UTILITY FUNCS ////////////////////////////////////////////////////////////////

GNODE *newGNODE(void *value) {
	GNODE *newNode = malloc(sizeof(GNODE));
	newNode->v = value;
	newNode->f = 1;
	return newNode;
}

GST *newGST(int(*c)(void *, void *)) {
	GST *newTree     = malloc(sizeof(GST));
	newTree->tree    = newBST(compareGNODE);
	setBSTdisplay(newTree->tree, displayGNODE);
	setBSTfree(newTree->tree, freeGNODE);
	newTree->display = 0;
	newTree->free    = 0;
	newTree->compare = c;
	newTree->swap    = swapper;
	newTree->d       = 0;
	return newTree;
}

void setGSTdisplay(GST *t, void(*d)(void *v, FILE *fp)) {
	t->display = d;
	return;
}

void setGSTswapper(GST *t, void(*s)(TNODE *, TNODE *)) {
	t->swap = s;
	setBSTswapper(t->tree, s);
	return;
}

void setGSTfree(GST *t, void(*f)(void *)) {
	t->free = f;
	return;
}

TNODE *getGSTroot(GST *t) {
	return getBSTroot(t->tree);
}

void setGSTroot(GST *t, TNODE *replacement) {
	setBSTroot(t->tree, replacement);
	return;
}
void setGSTsize(GST *t, int s) {
	setBSTsize(t->tree, s);
	return;
}

TNODE *insertGST(GST *t, void *value) {
	GNODE *newNode   = newGNODE(value);
	newNode->display = t->display;
	newNode->compare = t->compare;
	newNode->free    = t->free;

	if (findBST(t->tree, newNode)) {
		TNODE *temp = locateGST(t, newNode);
		GNODE *unwrap = unwrapGST(temp);
		unwrap->f++;
		t->d++;
		freeGNODE(newNode);
		return temp;
	}

	return insertBST(t->tree, newNode);
}

void *findGST(GST *t, void *key) {
	return unwrapGST(findBST(t->tree, key));
}

TNODE *locateGST(GST *t, void *key) {
	return locateBST(t->tree, key);
}

int deleteGST(GST *t, void *key) {
	GNODE *temp = unwrapGST(locateGST(t, key));
	if (temp == 0) {
		return -1;
	}
	if (temp->f > 1) {
		temp->f--;
		return temp->f;
	}
	else {
		return deleteBST(t->tree, key);
	}
	return -1;
}

TNODE *swapToLeafGST(GST *t, TNODE *node) {
	return swapToLeafBST(t->tree, node);
}

void pruneLeafGST(GST *t, TNODE *leaf) {
	return pruneLeafBST(t->tree, leaf);
}

int sizeGST(GST *t) {
	return sizeBST(t->tree);
}

void statisticsGST(GST *t, FILE *fp) {
	fprintf(fp, "Duplicates: %d\n", t->d);
	statisticsBST(t->tree, fp);
	return;
}

void displayGST(GST *t, FILE *fp) {
	displayBST(t->tree, fp);
	return;
}

void freeGST(GST *t) {
	freeBST(t->tree);
	free(t);
	return;
}

int debugGST(GST *t, int level) {
	return debugBST(t->tree, level);
}

void *unwrapGST(TNODE *node) {
	if (!node) {
		return NULL;
	}
	GNODE *temp = getTNODEvalue(node);
	return temp->v;
}

int freqGST(GST *g, void *key) {
	TNODE *temp = locateGST(g, key);
	if (temp) {
		GNODE *unwrap = getTNODEvalue(temp);
		return unwrap->f;
	}
	return 0;
}

int duplicatesGST(GST *g) {
	return g->d;
}