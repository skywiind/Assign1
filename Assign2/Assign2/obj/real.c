#include <stdio.h>
#include <stdlib.h>
#include <limits.h>
#include <assert.h>
#include "real.h"

struct REAL{
    double value;
};

REAL* newREAL(double x){
    REAL *p = malloc(sizeof(REAL));
    assert(p != 0);
    p->value = x;
    return p;
}

double getREAL(REAL *v){
    return v->value;
}

double setREAL(REAL *v,double x){
    double old = v->value;
    v->value = x;
    return old;
}

void displayREAL(void *v,FILE *fp){
    fprintf(fp,"%lf",getREAL((REAL*) v));
	return;
}

double compareREAL(void *v,void *w){
    return getREAL(v) - getREAL(w);
}

void freeREAL(void *v){
    free((REAL *) v);
	return;
}
