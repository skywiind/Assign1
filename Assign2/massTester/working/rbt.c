/*****************************************************************************************
*                                           *                                            *
*             CS201-001 Assign 2            * This file defines the implementations of   *
*             Wyatt Fairbanks               * the RBT class methods.                     *
*             Dr. John Lusth                *                                            *
*             rbt.c                         *                                            *
*             10/22/2018                    *                                            *
*                                           *                                            *
*****************************************************************************************/
#include <stdio.h>
#include <stdlib.h>
#include "gst.h"
#include "rbt.h"

//RED   = 0
//BLACK = 1

struct rbt {
	GST   *tree;
	void (*display) (void *v, FILE *fp);
	void (*free) (void *v);
	void (*swap) (TNODE *v1, TNODE* v2);
	int  (*compare) (void *v1, void *v2);
};

typedef struct rbnode RBNODE;

struct rbnode {
	void *v;
	int   c;
	void(*display) (void *value, FILE *fp);
	void(*free) (void *value);
	int(*compare) (void *v1, void *v2);
};

//UTILITY FUNCTIONS ////////////////////////////////////////////////////////////////

void displayRBNODE(void *v, FILE *fp) {
	RBNODE *temp = v;
	if (temp == NULL) {
		return;
	}
	temp->display(temp->v, fp);
	if (temp->c == 0) {
		fprintf(fp, "*");
	}
	return;
}

int compareRBNODE(void *v, void *w) {
	RBNODE *a = v;
	RBNODE *b = w;
	return  a->compare(a->v, b->v);
}

int compareShitstorm(void *v, void *w) {
	RBNODE *a = v;
	RBNODE *b = w;
	return  a->compare(a->v, b);
}

void freeRBNODE(void *n) {
	RBNODE *temp = n;
	temp->free(temp->v);
	free(temp);
	return;
}

void RBTswapper(TNODE *a, TNODE *b) {
	void *va = getTNODEvalue(a);   //get the GST value
	void *vb = getTNODEvalue(b);   //get the GST value
	// swap the GST values
	setTNODEvalue(a, vb);
	setTNODEvalue(b, va);
	// the above swap swapped the colors,
	// but the colors should stay with the nodes,
	// so swap the colors back to the original nodes
	RBNODE *x = unwrapGST(a);
	RBNODE *y = unwrapGST(b);
	int color = x->c;
	x->c = y->c;
	y->c = color;
	return;
}

void setNodeColor(TNODE *n, int color) {
	if (!n) {
		return;
	}
	RBNODE *temp = unwrapGST(n);
	temp->c = color;
	return;
}

int getNodeColor(TNODE *n) {
	if (!n) {
		return -1;
	}
	RBNODE *temp = unwrapGST(n);
	return temp->c;
}

int isNodeLeft(TNODE *n) {
	if (!n) {
		return -1;
	}
	if (getTNODEleft(getTNODEparent(n)) == n) {
		return 1;
	}
	return 0;
}

int isNodeRight(TNODE *n) {
	if (!n) {
		return -1;
	}
	if (getTNODEright(getTNODEparent(n)) == n) {
		return 1;
	}
	return 0;
}

TNODE *getGrandparent(TNODE *n) {
	if (!n) {
		return NULL;
	}
	return getTNODEparent(getTNODEparent(n));
}

TNODE *getUncle(TNODE *n) {
	if (!n) {
		return NULL;
	}
	//Is the parent node a right or left child?
	if (getTNODEright(getGrandparent(n)) == getTNODEparent(n)) {
		return getTNODEleft(getGrandparent(n));
	}
	else if (getTNODEleft(getGrandparent(n)) == getTNODEparent(n)) {
		return getTNODEright(getGrandparent(n));
	}
	return NULL;
}
TNODE *getSibling(TNODE *n) {
	if (!n) {
		return NULL;
	}
	if (isNodeRight(n) == 1) {
		return getTNODEleft(getTNODEparent(n));
	}
	else {
		return getTNODEright(getTNODEparent(n));
	}
}

TNODE *getNephew(TNODE *n) {
	if (!n) {
		return NULL;
	}
	if (isNodeRight(n) == 1) {
		return getTNODEleft(getSibling(n));
	}
	else {
		return getTNODEright(getSibling(n));
	}
}

TNODE *getNiece(TNODE *n) {
	if (!n) {
		return NULL;
	}
	if (isNodeRight(n) == 1) {
		return getTNODEright(getSibling(n));
	}
	else {
		return getTNODEleft(getSibling(n));
	}
}

void rotateLeft(RBT *t, TNODE *b) {
	TNODE *a = NULL;
	TNODE *oldLeft = NULL;
	a = getTNODEparent(b);
	setTNODEparent(b, getTNODEparent(a));
	if (getRBTroot(t) != a) {
		if (isNodeLeft(a)) {
			setTNODEleft(getTNODEparent(a), b);
		}
		else {
			setTNODEright(getTNODEparent(a), b);
		}
	}

	oldLeft = getTNODEleft(b);
	setTNODEleft(b, a);
	setTNODEparent(a, b);
	setTNODEright(a, oldLeft);
	if (oldLeft != NULL) {
		setTNODEparent(oldLeft, a);
	}
	if (getRBTroot(t) == a) {
		setRBTroot(t, b);
		setTNODEparent(b, b);
	}
	return;
}

void rotateRight(RBT *t, TNODE *b) {
	TNODE *a = NULL;
	TNODE *oldRight = NULL;
	a = getTNODEparent(b);
	setTNODEparent(b, getTNODEparent(a));
	if (getRBTroot(t) != a) {
		if (isNodeLeft(a)) {
			setTNODEleft(getTNODEparent(a), b);
		}
		else {
			setTNODEright(getTNODEparent(a), b);
		}
	}

	oldRight = getTNODEright(b);
	setTNODEright(b, a);
	setTNODEparent(a, b);
	setTNODEleft(a, oldRight);
	if (oldRight != NULL) {
		setTNODEparent(oldRight, a);
	}
	if (getRBTroot(t) == a) {
		setRBTroot(t, b);
		setTNODEparent(b, b);
	}
	return;
}
/*
void rotateLeft(RBT *t, TNODE* x) {
	if (!x) {
		return;
	}
	int l = 0;
	if (!getTNODEright(x) && !getTNODEleft(x)) {
		l = 1;
	}
	TNODE *y = getTNODEright(x);
	setTNODEright(x, getTNODEleft(y));
	if (getTNODEleft(y)) {
		setTNODEparent(getTNODEright(x), x);
	}
	setTNODEparent(y, getTNODEparent(x));
	if (getTNODEparent(x) == x) {
		setRBTroot(t, y);
	}
	else if (isNodeLeft(x)) {
		setTNODEleft(getTNODEparent(x), y);
	}
	else {
		setTNODEright(getTNODEparent(x), y);
	}
	setTNODEleft(y, x);
	setTNODEparent(x, y);
	if (l != 0) {
		if (!getTNODEright(x) && !getTNODEleft(x)) {
			return;
		}
		else {
			printf("Error in rotation.\n");
		}
	}
	return;
}


void rotateRight(RBT *t, TNODE *x) {
	if (!x) {
		return;
	}
	int l = 0;
	TNODE *y = getTNODEleft(x);
	if(!getTNODEright(x) && !getTNODEleft(x)) {
		l = 1;
	}
	setTNODEleft(x, getTNODEright(y));
	if (getTNODEright(y)) {
		setTNODEparent(getTNODEleft(x), x);
	}
	setTNODEparent(y, getTNODEparent(x));
	if (getTNODEparent(x) == x) {
		setRBTroot(t, y);
	}
	else if (isNodeLeft(x)) {
		setTNODEleft(getTNODEparent(x), y);
	}
	else {
		setTNODEright(getTNODEparent(x), y);
	}
	setTNODEright(y, x);
	setTNODEparent(x, y);
	if (l != 0) {
		if (!getTNODEright(x) && !getTNODEleft(x)) {
			return;
		}
		else {
			printf("Error in rotation.\n");
		}
	}
	return;
}*/

void insertFixup(RBT *t, TNODE *inserted) {
	TNODE *x = inserted;
	//while x is not root
	while (getRBTroot(t) != x) {
		//if the parent is black
		if (getNodeColor(getTNODEparent(x)) == 1) {
			break;
		}
		//if the uncle is red
		if (getNodeColor(getUncle(x)) == 0) {
			//make the parent black
			setNodeColor(getTNODEparent(x), 1);
			//make the uncle black
			setNodeColor(getUncle(x), 1);
			//make the grandparent red
			setNodeColor(getGrandparent(x), 0);
			x = getGrandparent(x);
		}
		//the uncle is black
		else {
			//is the node the left child of a right child
			if (isNodeLeft(x)) {
				if (isNodeRight(getTNODEparent(x))) {
					rotateRight(t, x);
					x = getTNODEright(x);
				}
			}
			//is the node the right child of a left child
			else if (isNodeRight(x)) {
				if (isNodeLeft(getTNODEparent(x))) {
					rotateLeft(t, x);
					x = getTNODEleft(x);
				}
			}
			//make the parent black
			setNodeColor(getTNODEparent(x), 1);
			//make the grandparent red
			setNodeColor(getGrandparent(x), 0);
			//is the parent a left child
			if (isNodeLeft(getTNODEparent(x)) == 1) {
				rotateRight(t, getTNODEparent(x));
			}
			else {
				rotateLeft(t, getTNODEparent(x));
			}
			break;
		}
	}
	//recolor the root 
	setNodeColor(getRBTroot(t), 1);
	return;
}

void deleteFixup(RBT *t, TNODE *deleted) {
	TNODE *x = deleted;
	while (getRBTroot(t) != x) {
		if (getNodeColor(x) == 0) {
			break;
		}
		if (getNodeColor(getSibling(x)) == 0) {
			setNodeColor(getTNODEparent(x), 0);
			setNodeColor(getSibling(x), 0);
			if (isNodeLeft(getSibling(x)) == 1) {
				rotateRight(t, getSibling(x));
			}
			else {
				rotateLeft(t, getSibling(x));
			}
		}
		else if (getNodeColor(getNephew(x)) == 0) {
			setNodeColor(getSibling(x), getNodeColor(getTNODEparent(x)));
			setNodeColor(getTNODEparent(x), 1);
			setNodeColor(getNephew(x), 1);
			if (isNodeLeft(getSibling(x)) == 1) {
				rotateRight(t, getSibling(x));
			}
			else {
				rotateLeft(t, getSibling(x));
			}
			break;
		}
		else if (getNodeColor(getNiece(x)) == 0) {
			setNodeColor(getNiece(x), 1);
			setNodeColor(getSibling(x), 0);
			if (isNodeLeft(getNiece(x)) == 1) {
				rotateRight(t, getNiece(x));
			}
			else {
				rotateLeft(t, getSibling(x));
			}
		}
		else {
			setNodeColor(getSibling(x), 0);
			x = getTNODEparent(x);
		}
	}
	setNodeColor(x, 1);
	return;
}

//END UTILITY FUNCS ////////////////////////////////////////////////////////////////

RBNODE *newRBNODE(void *value) {
	RBNODE *newNode = malloc(sizeof(RBNODE));
	newNode->v = value;
	newNode->c = 0;
	return newNode;
}

RBT *newRBT(int(*c)(void *, void*)) {
	RBT *newTree     = malloc(sizeof(RBT));
	newTree->tree    = newGST(compareRBNODE);
	setGSTdisplay(newTree->tree, displayRBNODE);
	setGSTfree(newTree->tree, freeRBNODE);
	newTree->display = 0;
	newTree->free    = 0;
	newTree->compare = c;
	newTree->swap    = RBTswapper;
	return newTree;
}

void setRBTdisplay(RBT *t, void(*d)(void *v, FILE *fp)) {
	t->display = d;
	return;
}

void setRBTswapper(RBT *t, void(*s)(TNODE*, TNODE*)) {
	t->swap = s;
	setGSTswapper(t->tree, s);
	return;
}

void setRBTfree(RBT *t, void(*f)(void*)) {
	t->free = f;
	return;
}

TNODE *getRBTroot(RBT *t) {
	return getGSTroot(t->tree);
}

void setRBTroot(RBT *t, TNODE *replacement) {
	setGSTroot(t->tree, replacement);
	return;
}

void setRBTsize(RBT *t, int s) {
	setGSTsize(t->tree, s);
	return;
}

TNODE *insertRBT(RBT *t, void *v) {
	int found = 0;
	TNODE* inserted = NULL;
	RBNODE *newNode  = newRBNODE(v);
	newNode->display = t->display;
	newNode->compare = compareShitstorm;
	newNode->free    = t->free;
	if (findGST(t->tree, newNode)) {
		found = 1;
	}
	
	newNode->compare = t->compare;
	inserted = insertGST(t->tree, newNode);

	if (!found) {
		insertFixup(t, inserted);
	}
	return inserted;
}

void *findRBT(RBT *t, void *key) {
	RBNODE *temp  = newRBNODE(key);
	temp->display = t->display;
	temp->compare = compareShitstorm;
	temp->free    = t->free;
	void *ans     = findGST(t->tree, temp);
	free(temp);
	return ans;
}

TNODE *locateRBT(RBT *t, void *key) {
	RBNODE *temp  = newRBNODE(key);
	temp->display = t->display;
	temp->compare = compareShitstorm;
	temp->free    = t->free;
	TNODE *ans    = locateGST(t->tree, temp);
	free(temp);
	return ans;
}

int deleteRBT(RBT *t, void *key) {
	TNODE *temp = locateRBT(t, key);
	if (!temp) {
		return -1;
	}
	int f = freqRBT(t, key);
	RBNODE *unwrap = unwrapGST(temp);
	if (f > 1) {
		deleteGST(t->tree, unwrap);
		return f - 1;
		/*RBNODE *target = newRBNODE(key);
		int ans = deleteGST(t->tree, target);
		free(target);
		return ans;*/
	}
	else {
		int size = sizeRBT(t);
		TNODE *node = swapToLeafGST(t->tree, temp);
		deleteFixup(t, node);
		pruneLeafGST(t->tree, node);
		setRBTsize(t, size - 1);
		return 0;
	}
	return -1;
}

TNODE *swapToLeafRBT(RBT *t, TNODE *node) {
	return swapToLeafGST(t->tree, node);
}

void pruneLeafRBT(RBT *t, TNODE *leaf) {
	return pruneLeafGST(t->tree, leaf);
}

int sizeRBT(RBT *t) {
	return sizeGST(t->tree);
}

void statisticsRBT(RBT *t, FILE *fp) {
	return statisticsGST(t->tree, fp);
}

void displayRBT(RBT *t, FILE *fp) {
	return displayGST(t->tree, fp);
}

void freeRBT(RBT *t) {
	freeGST(t->tree);
	free(t);
	return;
}

int debugRBT(RBT *t, int level) {
	return debugGST(t->tree, level);
}

void *unwrapRBT(TNODE *n) {
	if (!n) return NULL;
	RBNODE *temp = unwrapGST(n);
	return temp->v;
}

int freqRBT(RBT *g, void *key) {
	RBNODE *temp = newRBNODE(key);
	int freq = freqGST(g->tree, temp);
	free(temp);
	return freq;
}

/*int freqRBT(RBT *g, void *key) {
	TNODE *temp = locateRBT(g, key);
	if (temp) {
		RBNODE *unwrap = unwrapGST(temp);
	}
	return freqGST(g->tree, key);
}*/

int duplicatesRBT(RBT *g) {
	return duplicatesGST(g->tree);
}