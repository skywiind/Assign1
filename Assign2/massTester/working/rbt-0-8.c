#include <stdio.h>
#include <stdlib.h>
#include "gst.h"
#include "rbt.h"
#include "integer.h"
#include "real.h"
#include "string.h"

void srandom(unsigned int);
long int random(void);

int
main(void)
    {
    srandom(20);
    printf("INTEGER test of RBT, mixed insertions and deletions\n");
    int i;
    RBT *p = newRBT(compareINTEGER);
    setRBTdisplay(p,displayINTEGER);
    setRBTfree(p,freeINTEGER);
    for (i = 0; i < 200000; ++i)
        {
        int j = random() % 33333;
        INTEGER *a = newINTEGER(j);
        if (random() % 10 < 6)
            {
            if (freqRBT(p,a) > 1)
                deleteRBT(p,a);
            else
                {
                INTEGER *x = findRBT(p,a);
                setRBTfree(p,0);
                deleteRBT(p,a);
                setRBTfree(p,freeINTEGER);
                freeINTEGER(x);
                }
            freeINTEGER(a);
            }
        else
             insertRBT(p,a);
        }
    if (sizeRBT(p) < 200)
        {
        debugRBT(p,0);
        printf("RBT:\n");
        displayRBT(p,stdout);
        debugRBT(p,1);
        printf("debug (in-order): ");
        displayRBT(p,stdout);
        printf("\n");
        debugRBT(p,2);
        printf("debug (pre-order): ");
        displayRBT(p,stdout);
        printf("\n");
        debugRBT(p,3);
        printf("debug (post-order): ");
        displayRBT(p,stdout);
        printf("\n");
        }
    printf("size is %d\n",sizeRBT(p));
    statisticsRBT(p,stdout);
    while (sizeRBT(p) > 200)
        {
        for (i = 0; i < 100000; ++i)
            {
            int j = random() % 50000;
            INTEGER *a = newINTEGER(j);
            if (freqRBT(p,a) > 1)
                deleteRBT(p,a);
            else
                {
                INTEGER *x = findRBT(p,a);
                setRBTfree(p,0);
                deleteRBT(p,a);
                setRBTfree(p,freeINTEGER);
                freeINTEGER(x);
                }
            freeINTEGER(a);
            }
        }
    if (sizeRBT(p) < 200)
        {
        debugRBT(p,0);
        printf("RBT:\n");
        displayRBT(p,stdout);
        debugRBT(p,1);
        printf("debug: ");
        displayRBT(p,stdout);
        printf("\n");
        debugRBT(p,2);
        printf("debug: ");
        displayRBT(p,stdout);
        printf("\n");
        debugRBT(p,3);
        printf("debug: ");
        displayRBT(p,stdout);
        printf("\n");
        }
    printf("size is %d\n",sizeRBT(p));
    statisticsRBT(p,stdout);
    setRBTfree(p,freeINTEGER);
    freeRBT(p);
    return 0;
    }
