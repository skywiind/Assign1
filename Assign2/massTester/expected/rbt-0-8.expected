INTEGER test of RBT, mixed insertions and deletions
size is 17928
Duplicates: 14406
Nodes: 17928
Minimum depth: 10
Maximum depth: 17
RBT:
0: 13134(13134)X
1: 7372(13134)L 20105<2>(13134)R
2: 3976(7372)L 9708<2>(7372)R 16369(20105<2>)L 25816*<2>(20105<2>)R
3: 1430<3>(3976)L 6232*(3976)R 8263*<3>(9708<2>)L 11686(9708<2>)R 15252(16369)L 18463<3>(16369)R 22949<2>(25816*<2>)L 29557(25816*<2>)R
4: 1103*(1430<3>)L 3521(1430<3>)R 4576(6232*)L 6933(6232*)R 7932<2>(8263*<3>)L 8716(8263*<3>)R 10637(11686)L 12356(11686)R 14714(15252)L 15903*<2>(15252)R 17940*(18463<3>)L 19458(18463<3>)R 21354<2>(22949<2>)L 24562<2>(22949<2>)R 28237(29557)L 30617(29557)R
5: 603<2>(1103*)L 1314(1103*)R 2369*(3521)L 3708(3521)R 4123<2>(4576)L 5417(4576)R 6693(6933)L 6970(6933)R 7606(7932<2>)L 8174(7932<2>)R 8683<2>(8716)L 9218(8716)R =10257(10637)L 11413*<2>(10637)R 12246<2>(12356)L 13030(12356)R 14283(14714)L =14935(14714)R 15651<3>(15903*<2>)L 16139<2>(15903*<2>)R 16943<3>(17940*)L 18080<2>(17940*)R =19434(19458)L 19740*(19458)R 20733<2>(21354<2>)L 22403*(21354<2>)R 23700(24562<2>)L 25685<4>(24562<2>)R 27139(28237)L 28821<2>(28237)R 30172<2>(30617)L 31671(30617)R
6: 540<2>(603<2>)L =975<2>(603<2>)R =1295(1314)L =1372<4>(1314)R =2161(2369*)L 2693(2369*)R =3599*<2>(3708)L =4118(4123<2>)L 4449<2>(4123<2>)R =5096(5417)L 6090*<2>(5417)R =6634(6693)L =6702<2>(6693)R =6959(6970)L =7198<5>(6970)R =7529<2>(7606)L 7665<3>(7606)R =8150(8174)L =8245(8174)R =8461<2>(8683<2>)L =8692<2>(8683<2>)R 9110<2>(9218)L 9477*(9218)R =10998<3>(11413*<2>)L =11576(11413*<2>)R =12063*<3>(12246<2>)L =12878*(13030)L =14447*<2>(14283)R =15510(15651<3>)L =15855<4>(15651<3>)R =16056(16139<2>)L =16231<2>(16139<2>)R 16711(16943<3>)L 17342<2>(16943<3>)R =18050(18080<2>)L =18193(18080<2>)R =19689(19740*)L =19900(19740*)R 20461(20733<2>)L 20811*<3>(20733<2>)R 21998(22403*)L 22772(22403*)R =23411(23700)L 23885*(23700)R =25430(25685<4>)L =25748(25685<4>)R 26199(27139)L =27984(27139)R 28561*(28821<2>)L 29255(28821<2>)R 29894(30172<2>)L 30590*<2>(30172<2>)R =31180(31671)L =32225(31671)R
7: =133*<2>(540<2>)L =2675*(2693)L =4415*<2>(4449<2>)L =5728<2>(6090*<2>)L =6145<2>(6090*<2>)R =7659*(7665<3>)L =9173*<2>(9110<2>)R 9466(9477*)L =9606(9477*)R =16569*(16711)L =17904*(17342<2>)R =20672*(20461)R =20769(20811*<3>)L 21259(20811*<3>)R 21695*<2>(21998)L =22254<2>(21998)R =22726(22772)L =22913(22772)R =23835(23885*)L =24402(23885*)R =27030*(26199)R =28546(28561*)L =28713<2>(28561*)R =29234*(29255)L =30115*(29894)R =30340(30590*<2>)L =30594<3>(30590*<2>)R
8: =9469*<2>(9466)R =21324*<3>(21259)R =21617(21695*<2>)L =21725<2>(21695*<2>)R
debug: [[[[[[[[133*<2>] 540<2>] 603<2> [975<2>]] 1103* [[1295] 1314 [1372<4>]]] 1430<3> [[[2161] 2369* [[2675*] 2693]] 3521 [[3599*<2>] 3708]]] 3976 [[[[4118] 4123<2> [[4415*<2>] 4449<2>]] 4576 [[5096] 5417 [[5728<2>] 6090*<2> [6145<2>]]]] 6232* [[[6634] 6693 [6702<2>]] 6933 [[6959] 6970 [7198<5>]]]]] 7372 [[[[[7529<2>] 7606 [[7659*] 7665<3>]] 7932<2> [[8150] 8174 [8245]]] 8263*<3> [[[8461<2>] 8683<2> [8692<2>]] 8716 [[9110<2> [9173*<2>]] 9218 [[9466 [9469*<2>]] 9477* [9606]]]]] 9708<2> [[[10257] 10637 [[10998<3>] 11413*<2> [11576]]] 11686 [[[12063*<3>] 12246<2>] 12356 [[12878*] 13030]]]]] 13134 [[[[[14283 [14447*<2>]] 14714 [14935]] 15252 [[[15510] 15651<3> [15855<4>]] 15903*<2> [[16056] 16139<2> [16231<2>]]]] 16369 [[[[[16569*] 16711] 16943<3> [17342<2> [17904*]]] 17940* [[18050] 18080<2> [18193]]] 18463<3> [[19434] 19458 [[19689] 19740* [19900]]]]] 20105<2> [[[[[20461 [20672*]] 20733<2> [[20769] 20811*<3> [21259 [21324*<3>]]]] 21354<2> [[[[21617] 21695*<2> [21725<2>]] 21998 [22254<2>]] 22403* [[22726] 22772 [22913]]]] 22949<2> [[[23411] 23700 [[23835] 23885* [24402]]] 24562<2> [[25430] 25685<4> [25748]]]] 25816*<2> [[[[26199 [27030*]] 27139 [27984]] 28237 [[[28546] 28561* [28713<2>]] 28821<2> [[29234*] 29255]]] 29557 [[[29894 [30115*]] 30172<2> [[30340] 30590*<2> [30594<3>]]] 30617 [[31180] 31671 [32225]]]]]]]
debug: [13134 [7372 [3976 [1430<3> [1103* [603<2> [540<2> [133*<2>]] [975<2>]] [1314 [1295] [1372<4>]]] [3521 [2369* [2161] [2693 [2675*]]] [3708 [3599*<2>]]]] [6232* [4576 [4123<2> [4118] [4449<2> [4415*<2>]]] [5417 [5096] [6090*<2> [5728<2>] [6145<2>]]]] [6933 [6693 [6634] [6702<2>]] [6970 [6959] [7198<5>]]]]] [9708<2> [8263*<3> [7932<2> [7606 [7529<2>] [7665<3> [7659*]]] [8174 [8150] [8245]]] [8716 [8683<2> [8461<2>] [8692<2>]] [9218 [9110<2> [9173*<2>]] [9477* [9466 [9469*<2>]] [9606]]]]] [11686 [10637 [10257] [11413*<2> [10998<3>] [11576]]] [12356 [12246<2> [12063*<3>]] [13030 [12878*]]]]]] [20105<2> [16369 [15252 [14714 [14283 [14447*<2>]] [14935]] [15903*<2> [15651<3> [15510] [15855<4>]] [16139<2> [16056] [16231<2>]]]] [18463<3> [17940* [16943<3> [16711 [16569*]] [17342<2> [17904*]]] [18080<2> [18050] [18193]]] [19458 [19434] [19740* [19689] [19900]]]]] [25816*<2> [22949<2> [21354<2> [20733<2> [20461 [20672*]] [20811*<3> [20769] [21259 [21324*<3>]]]] [22403* [21998 [21695*<2> [21617] [21725<2>]] [22254<2>]] [22772 [22726] [22913]]]] [24562<2> [23700 [23411] [23885* [23835] [24402]]] [25685<4> [25430] [25748]]]] [29557 [28237 [27139 [26199 [27030*]] [27984]] [28821<2> [28561* [28546] [28713<2>]] [29255 [29234*]]]] [30617 [30172<2> [29894 [30115*]] [30590*<2> [30340] [30594<3>]]] [31671 [31180] [32225]]]]]]]
debug: [[[[[[[[133*<2>] 540<2>] [975<2>] 603<2>] [[1295] [1372<4>] 1314] 1103*] [[[2161] [[2675*] 2693] 2369*] [[3599*<2>] 3708] 3521] 1430<3>] [[[[4118] [[4415*<2>] 4449<2>] 4123<2>] [[5096] [[5728<2>] [6145<2>] 6090*<2>] 5417] 4576] [[[6634] [6702<2>] 6693] [[6959] [7198<5>] 6970] 6933] 6232*] 3976] [[[[[7529<2>] [[7659*] 7665<3>] 7606] [[8150] [8245] 8174] 7932<2>] [[[8461<2>] [8692<2>] 8683<2>] [[[9173*<2>] 9110<2>] [[[9469*<2>] 9466] [9606] 9477*] 9218] 8716] 8263*<3>] [[[10257] [[10998<3>] [11576] 11413*<2>] 10637] [[[12063*<3>] 12246<2>] [[12878*] 13030] 12356] 11686] 9708<2>] 7372] [[[[[[14447*<2>] 14283] [14935] 14714] [[[15510] [15855<4>] 15651<3>] [[16056] [16231<2>] 16139<2>] 15903*<2>] 15252] [[[[[16569*] 16711] [[17904*] 17342<2>] 16943<3>] [[18050] [18193] 18080<2>] 17940*] [[19434] [[19689] [19900] 19740*] 19458] 18463<3>] 16369] [[[[[[20672*] 20461] [[20769] [[21324*<3>] 21259] 20811*<3>] 20733<2>] [[[[21617] [21725<2>] 21695*<2>] [22254<2>] 21998] [[22726] [22913] 22772] 22403*] 21354<2>] [[[23411] [[23835] [24402] 23885*] 23700] [[25430] [25748] 25685<4>] 24562<2>] 22949<2>] [[[[[27030*] 26199] [27984] 27139] [[[28546] [28713<2>] 28561*] [[29234*] 29255] 28821<2>] 28237] [[[[30115*] 29894] [[30340] [30594<3>] 30590*<2>] 30172<2>] [[31180] [32225] 31671] 30617] 29557] 25816*<2>] 20105<2>] 13134]
size is 148
Duplicates: 77
Nodes: 148
Minimum depth: 5
Maximum depth: 8
